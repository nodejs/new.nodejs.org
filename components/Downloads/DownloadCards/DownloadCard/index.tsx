import classNames from 'classnames';
import { MdGetApp } from 'react-icons/md';
import { Tab } from 'react-tabs';
import Image from 'next/image';
import styles from './index.module.scss';
import { useRouter } from '@/hooks/useRouter';
import type { FC } from 'react';
import type { UserOS } from '@/types/userOS';

type DownloadCardProps = {
  os: UserOS;
  icon: string;
  label: string;
  download: string;
  filename: string;
  selected: boolean;
  onSelect: (os: string) => void;
};

const DownloadCard: FC<DownloadCardProps> = ({
  os,
  icon,
  label,
  download,
  filename,
  selected,
  onSelect,
}) => {
  const { basePath } = useRouter();

  const tabClassNames = classNames(styles.downloadCard, {
    [styles.downloadCardActive]: selected,
  });
  const selectCard = () => onSelect(os);
  const size = selected ? 56 : 48;

  return (
    <Tab
      className={tabClassNames}
      id={`download-card-${os}`}
      key={os}
      onClick={selectCard}
      selectedClassName={styles.downloadCardActive}
      tabIndex="0"
    >
      <div className={styles.top}>
        <Image
          height={size}
          width={size}
          src={`${basePath}/static/images/logos/${icon}`}
          alt={label}
        />
        {selected && (
          <a className={styles.link} href={download}>
            <MdGetApp />
          </a>
        )}
      </div>
      <p className={styles.label}>{label}</p>
      <p className={styles.filename} data-testid="lts">
        {filename}
      </p>
    </Tab>
  );
};

export default DownloadCard;
