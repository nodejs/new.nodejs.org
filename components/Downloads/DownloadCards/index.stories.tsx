import DownloadCards from './index';
import type { Meta as MetaObj, StoryObj } from '@storybook/react';

type Story = StoryObj<typeof DownloadCards>;
type Meta = MetaObj<typeof DownloadCards>;

export const Default: Story = {
  args: {
    versionWithPrefix: 'v18.15.0',
  },
};

export default { component: DownloadCards } as Meta;
