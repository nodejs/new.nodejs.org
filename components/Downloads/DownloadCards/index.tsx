import { Tabs, TabList } from 'react-tabs';
import { useState, useEffect } from 'react';
import styles from './index.module.scss';
import DownloadCard from './DownloadCard/index';
import type { FC, KeyboardEvent } from 'react';
import type { UserOS } from '../../../types/userOS';

type DownloadCardsProps = {
  version: string;
  userOS: UserOS;
};

const DownloadCards: FC<DownloadCardsProps> = ({ version, userOS }) => {
  const [selected, setSelected] = useState('');

  useEffect(() => {
    setSelected(!['WIN', 'MAC'].includes(userOS) ? 'SOURCECODE' : userOS);
  }, [userOS]);

  const downloadTypes = [
    {
      name: 'WIN',
      icon: 'windows-download-logo.svg',
      label: 'Windows Installer',
      download: `https://nodejs.org/dist/${version}/node-${version}-x86.msi`,
      filename: `node-${version}-x86.msi`,
    },
    {
      name: 'MAC',
      icon: 'mac-download-logo.svg',
      label: 'MAC Installer',
      download: `https://nodejs.org/dist/${version}/node-${version}.pkg`,
      filename: `node-${version}.pkg`,
    },
    {
      name: 'SOURCECODE',
      icon: 'source-code-download-logo.svg',
      label: 'Source Code',
      download: `https://nodejs.org/dist/${version}/node-${version}.tar.gz`,
      filename: `node-${version}.tar.gz`,
    },
  ];

  return (
    <Tabs tabIndex={-1} className={styles.downloadCardsWrapper}>
      <TabList
        className={styles.downloadCards}
        role="tablist"
        onKeyDown={(e: KeyboardEvent) => {
          const current = downloadTypes.findIndex(d => d.name === selected);

          let step = 0;
          if (e.key === 'ArrowLeft') {
            step = -1;
          } else if (e.key === 'ArrowRight') {
            step = 1;
          }
          if (!step) return;

          const next =
            (current + step + downloadTypes.length) % downloadTypes.length;
          setSelected(downloadTypes[next].name);
        }}
      >
        {downloadTypes.map(d => (
          <DownloadCard
            key={d.name}
            name={d.name}
            icon={d.icon}
            label={d.label}
            download={d.download}
            filename={d.filename}
            selected={selected === d.name}
            onSelect={setSelected}
          />
        ))}
      </TabList>
    </Tabs>
  );
};

export default DownloadCards;
