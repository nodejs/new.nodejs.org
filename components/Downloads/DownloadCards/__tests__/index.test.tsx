import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import DownloadCards from './../index';

jest.mock('next/router', () => ({
  useRouter: jest.fn().mockImplementation(() => ({
    basePath: '',
  })),
}));

describe('DownloadCards component', () => {
  it('renders correctly', () => {
    const { container } = render(
      <DownloadCards version="v18.15.0" userOS="MAC" />
    );
    expect(container).toMatchSnapshot();
  });

  it('tests click', async () => {
    render(<DownloadCards version="v18.15.0" userOS="MAC" />);
    const listElements = screen.getAllByRole('tab');

    expect(listElements.length).toEqual(3);
    expect(listElements[0]).toHaveClass('downloadCard');
    expect(listElements[1]).toHaveClass('downloadCard downloadCardActive');
    expect(listElements[2]).toHaveClass('downloadCard');

    await userEvent.click(listElements[0]);
    expect(listElements[0]).toHaveClass('downloadCard downloadCardActive');
    expect(listElements[1]).toHaveClass('downloadCard');
    expect(listElements[2]).toHaveClass('downloadCard');
  });

  it('tests keyboard events', () => {
    render(<DownloadCards version="v18.15.0" userOS="MAC" />);
    const tabListElement = screen.getByRole('tablist');
    const listElements = screen.getAllByRole('tab');

    expect(listElements.length).toEqual(3);
    expect(listElements[0]).toHaveClass('downloadCard');
    expect(listElements[1]).toHaveClass('downloadCard downloadCardActive');
    expect(listElements[2]).toHaveClass('downloadCard');

    // WIN <-MAC SOURCECODE
    fireEvent.keyDown(tabListElement, {
      key: 'ArrowLeft',
      code: 'ArrowLeft',
    });
    expect(listElements[0]).toHaveClass('downloadCard downloadCardActive');
    expect(listElements[1]).toHaveClass('downloadCard');
    expect(listElements[2]).toHaveClass('downloadCard');

    // <-WIN MAC SOURCECODE
    fireEvent.keyDown(tabListElement, {
      key: 'ArrowLeft',
      code: 'ArrowLeft',
    });
    expect(listElements[0]).toHaveClass('downloadCard');
    expect(listElements[1]).toHaveClass('downloadCard');
    expect(listElements[2]).toHaveClass('downloadCard downloadCardActive');

    // WIN MAC SOURCECODE->
    fireEvent.keyDown(tabListElement, {
      key: 'ArrowRight',
      code: 'ArrowRight',
    });
    expect(listElements[0]).toHaveClass('downloadCard downloadCardActive');
    expect(listElements[1]).toHaveClass('downloadCard');
    expect(listElements[2]).toHaveClass('downloadCard');
  });
});
