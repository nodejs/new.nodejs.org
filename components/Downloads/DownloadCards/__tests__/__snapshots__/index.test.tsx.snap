// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadCards component renders correctly 1`] = `
<div>
  <div
    class="downloadCardsWrapper"
    data-rttabs="true"
    tabindex="-1"
  >
    <ul
      class="downloadCards"
      role="tablist"
    >
      <li
        aria-controls="paneldownload-card-WIN"
        aria-disabled="false"
        aria-selected="false"
        class="downloadCard"
        data-rttab="true"
        id="tabdownload-card-WIN"
        role="tab"
        tabindex="0"
      >
        <div
          class="top"
        >
          <img
            alt="Windows Installer"
            data-nimg="1"
            decoding="async"
            height="48"
            loading="lazy"
            src="/static/images/logos/windows-download-logo.svg"
            style="color: transparent;"
            width="48"
          />
        </div>
        <p
          class="label"
        >
          Windows Installer
        </p>
        <p
          class="filename"
          data-testid="lts"
        >
          node-v18.15.0-x86.msi
        </p>
      </li>
      <li
        aria-controls="paneldownload-card-MAC"
        aria-disabled="false"
        aria-selected="false"
        class="downloadCard downloadCardActive"
        data-rttab="true"
        id="tabdownload-card-MAC"
        role="tab"
        tabindex="0"
      >
        <div
          class="top"
        >
          <img
            alt="MAC Installer"
            data-nimg="1"
            decoding="async"
            height="56"
            loading="lazy"
            src="/static/images/logos/mac-download-logo.svg"
            style="color: transparent;"
            width="56"
          />
          <a
            class="link"
            href="https://nodejs.org/dist/v18.15.0/node-v18.15.0.pkg"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"
              />
            </svg>
          </a>
        </div>
        <p
          class="label"
        >
          MAC Installer
        </p>
        <p
          class="filename"
          data-testid="lts"
        >
          node-v18.15.0.pkg
        </p>
      </li>
      <li
        aria-controls="paneldownload-card-SOURCECODE"
        aria-disabled="false"
        aria-selected="false"
        class="downloadCard"
        data-rttab="true"
        id="tabdownload-card-SOURCECODE"
        role="tab"
        tabindex="0"
      >
        <div
          class="top"
        >
          <img
            alt="Source Code"
            data-nimg="1"
            decoding="async"
            height="48"
            loading="lazy"
            src="/static/images/logos/source-code-download-logo.svg"
            style="color: transparent;"
            width="48"
          />
        </div>
        <p
          class="label"
        >
          Source Code
        </p>
        <p
          class="filename"
          data-testid="lts"
        >
          node-v18.15.0.tar.gz
        </p>
      </li>
    </ul>
  </div>
</div>
`;
